src/parser/jobs/nin/ShadowFang.js

```diff
@@ -0,0 +1,104 @@
+	}
+
+	_onDotApply(event) {
+		// Make sure we're tracking for this target
+		const lastApplication = this._lastApplication[event.targetID] = this._lastApplication[event.targetID] || 0
+
+		if (!lastApplication) {
+			this._lastApplication[event.targetID] = event.timestamp
+			return
+		}
+
+		// Base clip calc
+		let clip = SF_DURATION_MILLIS - (event.timestamp - lastApplication)
+
+		// Remove any untargetable time from the clip
+		clip -= this.invuln.getUntargetableUptime('all', event.timestamp - SF_DURATION_MILLIS, event.timestamp)
+
+		// Also remove invuln time in the future that casting later would just push dots into
+		// TODO: This relies on a full set of invuln data ahead of time. Can this be trusted?
+		clip -= this.invuln.getInvulnerableUptime('all', event.timestamp, event.timestamp + SF_DURATION_MILLIS + clip)
+
+		// Capping clip at 0 - less than that is downtime, which is handled by the checklist requirement
+		this._sfClip += Math.max(0, clip)
+
+		this._lastApplication[event.targetID] = event.timestamp
+	}
+
+	_onComplete() {
+		// Checklist rule for dot uptime
+		this.checklist.add(new Rule({
+			name: 'Keep Shadow Fang up',
+			description: <Fragment>
+				As a NIN, <ActionLink {...ACTIONS.SHADOW_FANG}/> is your strongest combo finisher (assuming at least 4 DoT ticks hit). In addition, it provides a slashing debuff which you, WARs, and SAMs are responsible for maintaining and should ideally never let lapse.
+			</Fragment>,
+			requirements: [
+				new Requirement({
+					name: <Fragment><ActionLink {...ACTIONS.SHADOW_FANG}/> uptime</Fragment>,
+					percent: () => this.getDotUptimePercent(),
+				}),
+			],
+		}))
+
+		// Suggestion for DoT clipping
+		const severity = this.getDotClippingSeverity()
+		if (severity) {
+			this.suggestions.add(new Suggestion({
+				icon: ACTIONS.SHADOW_FANG.icon,
+				content: <Fragment>
+					Avoid refreshing <ActionLink {...ACTIONS.SHADOW_FANG}/> significantly before its expiration, unless it would otherwise cost you significant uptime. Unnecessary refreshes risk overwriting buff snapshots and reduce the number of times you can use <ActionLink {...ACTIONS.AEOLIAN_EDGE}/>.
+				</Fragment>,
+				severity: severity,
+				why: <Fragment>
+					You lost {this.parser.formatDuration(this._sfClip)} of Shadow Fang to early refreshes.
+				</Fragment>,
+			}))
+		}
+	}
+
+	getDotUptimePercent() {
+		const statusUptime = this.enemies.getStatusUptime(STATUSES.SHADOW_FANG.id)
+		const fightDuration = this.parser.fightDuration - this.invuln.getInvulnerableUptime()
+		return (statusUptime / fightDuration) * 100
+	}
+
+	getDotClippingSeverity() {
+		// Simplified math, bitches
+		const fightDurationMillis = (this.parser.fightDuration - this.invuln.getInvulnerableUptime())
+		const clipSecsPerMin = Math.round((this._sfClip * 60) / fightDurationMillis)
+		return (clipSecsPerMin < 7 ? false : (clipSecsPerMin < 10 ? SEVERITY.MINOR : (clipSecsPerMin < 15 ? SEVERITY.MEDIUM : SEVERITY.MAJOR)))
```

Comment:
Check Mike's helper functions in `utilities` for future PRs, it'll simplify this a bunch.